File: sdk/windows/src/EndSessionClient.cs
// OIDC RP-initiated logout via POST /end_session

namespace Xid.Windows;

internal static class EndSessionClient
{
    internal static async Task EndSessionAsync(
        HttpClient http,
        Uri endSessionEndpoint,
        string idTokenHint,
        string? postLogoutRedirectUri = null,
        CancellationToken ct = default)
    {
        var form = new Dictionary<string, string>
        {
            ["id_token_hint"] = idTokenHint,
        };
        if (!string.IsNullOrEmpty(postLogoutRedirectUri))
            form["post_logout_redirect_uri"] = postLogoutRedirectUri;

        using var content = new FormUrlEncodedContent(form);
        using var request = new HttpRequestMessage(HttpMethod.Post, endSessionEndpoint) { Content = content };

        HttpResponseMessage response;
        try
        {
            response = await http.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            throw new XidException("无法访问 /end_session 端点。", "end_session_failed", ex);
        }

        if (!response.IsSuccessStatusCode)
            throw new XidException(
                $"/end_session 返回 HTTP {(int)response.StatusCode}。",
                "end_session_failed");
    }
}

File: sdk/ios/Sources/Xid/EndSessionClient.swift
// EndSessionClient.swift
// XID iOS Swift SDK
//
// OIDC RP-initiated logout via POST /end_session.

import Foundation

struct EndSessionClient: Sendable {
    private let urlSession: URLSession

    init(urlSession: URLSession = .shared) {
        self.urlSession = urlSession
    }

    /// POST id_token_hint (+ optional post_logout_redirect_uri) to /end_session.
    /// Best-effort: callers may ignore errors during sign-out.
    func endSession(
        endSessionEndpoint: URL,
        idTokenHint: String,
        postLogoutRedirectUri: URL? = nil
    ) async throws {
        var params: [String: String] = [
            "id_token_hint": idTokenHint,
        ]
        if let postLogoutRedirectUri {
            params["post_logout_redirect_uri"] = postLogoutRedirectUri.absoluteString
        }

        var request = URLRequest(url: endSessionEndpoint)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = params
            .map { "\($0.key)=\(urlEncode($0.value))" }
            .joined(separator: "&")
            .data(using: .utf8)

        let (_, response) = try await urlSession.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw XidError.endSessionFailed("无效响应类型")
        }
        guard (200 ... 399).contains(httpResponse.statusCode) else {
            throw XidError.endSessionFailed("HTTP \(httpResponse.statusCode)")
        }
    }

    private func urlEncode(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
    }
}